src/server/project.ts

```diff
@@ -874,52 +913,55 @@ namespace ts.server {
             super.setCompilerOptions(newOptions);
         }
 
-        // Used to keep track of what directories are watched for this project
-        directoriesWatchedForTsconfig: string[] = [];
-
-        constructor(projectService: ProjectService, documentRegistry: DocumentRegistry, compilerOptions: CompilerOptions, projectRootPath?: string) {
+        constructor(projectService: ProjectService, documentRegistry: DocumentRegistry, compilerOptions: CompilerOptions, public readonly projectRootPath?: string | undefined) {
             super(InferredProject.newName(),
                 ProjectKind.Inferred,
                 projectService,
                 documentRegistry,
                 /*files*/ undefined,
                 /*languageServiceEnabled*/ true,
                 compilerOptions,
-                /*compileOnSaveEnabled*/ false);
-            this.projectRootPath = projectRootPath;
+                /*compileOnSaveEnabled*/ false,
+                projectService.host);
         }
 
         addRoot(info: ScriptInfo) {
+            this.projectService.startWatchingConfigFilesForInferredProjectRoot(info);
             if (!this._isJsInferredProject && info.isJavaScript()) {
                 this.toggleJsInferredProject(/*isJsInferredProject*/ true);
             }
             super.addRoot(info);
         }
 
         removeRoot(info: ScriptInfo) {
+            this.projectService.stopWatchingConfigFilesForInferredProjectRoot(info, WatcherCloseReason.NotNeeded);
+            super.removeRoot(info);
             if (this._isJsInferredProject && info.isJavaScript()) {
-                if (filter(this.getRootScriptInfos(), info => info.isJavaScript()).length === 0) {
+                if (!some(this.getRootScriptInfos(), info => info.isJavaScript())) {

```

Comment:
Would `every(this.getRootScriptInfos(), info => !info.isJavaScript()))` be better?

Also rather scary to have the nearby `info` get shadowed by a lambda parameter's name - would prefer they be distinct